lib/channel.js

```diff
@@ -81,12 +82,29 @@ C.sendOrEnqueue = function(method, fields, reply) {
   }
 };
 
+C.waitForWrite = function () {
+  return new Promise((resolve) => {
+    var buffer = this.connection.channels[this.ch].buffer;
+    this.pendingWrites.push({resolve});
+
+    if (buffer.listenerCount('inbound_empty') === 0) {
+      buffer.once('inbound_empty', () => {
+        this.pendingWrites.forEach((promise) => {
+          promise.resolve();
+        });
+      });
+    }
+  });
+}
+
 C.sendMessage = function(fields, properties, content) {
-  return this.connection.sendMessage(
+  this.connection.sendMessage(
     this.ch,
     defs.BasicPublish, fields,
     defs.BasicProperties, properties,
     content);
+
+   return this.waitForWrite();
```

Comment:
I feel like there could be a race condition here if `sendMessage` is so absurdly fast that it completes before we can subscribe to the event. Maybe

```javascript
var promise = this.waitForWrite();

this.connection.sendMessage();

return promise;
```

?

Reply:
🤔

Say the passthrough being used is currently being written to rabbit.
What if you `waitForWrite();` and then `inbound_empty` is emitted before `this.connection.sendMessage();` has a chance to write more data to the passthrough?

The promise will resolve before it's truly "time".

I think the race condition you mention is less likely than the above, but yours is... More dangerous?

Reply:
:mind: